File: ios/monolens/Sources/monolens/MonolensPlugin.swift
import Flutter
import UIKit

/// The iOS end of the Pigeon bridge.
///
/// Nothing here does real work: it routes onto a background queue and hands off
/// to `MediaProbe`, `ImageTransformer` and `VideoExporter`. Decoding a 12 MP
/// still or exporting a clip on the platform thread would stall the very
/// viewfinder the author is looking at.
public class MonolensPlugin: NSObject, FlutterPlugin, MonolensHostApi {

  private var exporter: VideoExporter!
  private let queue = DispatchQueue(
    label: "com.monorithm.monolens.work", qos: .userInitiated, attributes: .concurrent)

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = MonolensPlugin()
    let messenger = registrar.messenger()

    let progress = MonolensFlutterApi(binaryMessenger: messenger)
    instance.exporter = VideoExporter { jobId, value in
      // Progress arrives on an export queue; the channel wants the main thread.
      DispatchQueue.main.async {
        progress.onProgress(jobId: jobId, progress: value) { _ in }
      }
    }

    MonolensHostApiSetup.setUp(binaryMessenger: messenger, api: instance)
    registrar.publish(instance)
  }

  // MARK: - MonolensHostApi

  func cacheDirectory() throws -> String {
    // Trailing-slash-free, because Dart joins with an explicit separator.
    (NSTemporaryDirectory() as NSString).standardizingPath
  }

  func probe(path: String, completion: @escaping (Result<MediaInfo, Error>) -> Void) {
    queue.async {
      do {
        Self.reply(completion, .success(try MediaProbe.probe(path: path)))
      } catch {
        Self.reply(completion, .failure(Self.mapped(error)))
      }
    }
  }

  func editImage(
    request: ImageEditRequest, completion: @escaping (Result<MediaInfo, Error>) -> Void
  ) {
    queue.async {
      do {
        Self.reply(completion, .success(try ImageTransformer.apply(request)))
      } catch {
        Self.reply(completion, .failure(Self.mapped(error)))
      }
    }
  }

  func trimVideo(
    request: VideoTrimRequest, jobId: String,
    completion: @escaping (Result<MediaInfo, Error>) -> Void
  ) {
    queue.async { [weak self] in
      guard let self else { return }
      self.exporter.trim(request: request, jobId: jobId) { result in
        Self.reply(completion, result.mapError(Self.mapped))
      }
    }
  }

  func cancel(jobId: String) throws {
    exporter.cancel(jobId: jobId)
  }

  func videoThumbnails(
    path: String, atMs: [Int64], maxDimension: Int64,
    completion: @escaping (Result<[FlutterStandardTypedData], Error>) -> Void
  ) {
    queue.async { [weak self] in
      guard let self else { return }
      do {
        let frames = try self.exporter.thumbnails(
          path: path, atMs: atMs, maxDimension: maxDimension)
        Self.reply(completion, .success(frames.map { FlutterStandardTypedData(bytes: $0) }))
      } catch {
        Self.reply(completion, .failure(Self.mapped(error)))
      }
    }
  }

  // MARK: - Plumbing

  /// Pigeon's reply handlers post onto the platform channel, which expects the
  /// main thread.
  private static func reply<T>(
    _ completion: @escaping (Result<T, Error>) -> Void, _ result: Result<T, Error>
  ) {
    DispatchQueue.main.async { completion(result) }
  }

  /// Turns a `MonolensError` into the `PigeonError` whose code Dart branches on
  /// (`monolens/cancelled` in particular). Anything else passes through.
  private static func mapped(_ error: Error) -> Error {
    (error as? MonolensError)?.pigeonError ?? error
  }
}

File: android/src/main/kotlin/com/monorithm/monolens/MonolensPlugin.kt
package com.monorithm.monolens

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * The Android end of the Pigeon bridge.
 *
 * Nothing here does real work: it routes onto a background executor and hands
 * off to [MediaProbe], [ImageTransformer] and [VideoExporter]. Decoding a 12 MP
 * still on the platform thread would stall the very viewfinder the author is
 * looking at. Trimming is the exception — Media3's Transformer wants a looper
 * thread and manages its own workers, so [VideoExporter] hops to main itself.
 */
class MonolensPlugin : FlutterPlugin, MonolensHostApi {

    private lateinit var context: Context
    private lateinit var exporter: VideoExporter
    private lateinit var progress: MonolensFlutterApi
    private lateinit var worker: ExecutorService

    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        worker = Executors.newCachedThreadPool()
        progress = MonolensFlutterApi(binding.binaryMessenger)
        exporter = VideoExporter(context) { jobId, value ->
            // Progress may arrive off the platform thread; the channel wants main.
            mainHandler.post { progress.onProgress(jobId, value) { } }
        }
        MonolensHostApi.setUp(binding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        MonolensHostApi.setUp(binding.binaryMessenger, null)
        worker.shutdown()
    }

    // MonolensHostApi

    override fun cacheDirectory(): String = context.cacheDir.absolutePath

    override fun probe(path: String, callback: (Result<MediaInfo>) -> Unit) {
        worker.execute { reply(callback) { MediaProbe.probe(path) } }
    }

    override fun editImage(
        request: ImageEditRequest,
        callback: (Result<MediaInfo>) -> Unit,
    ) {
        worker.execute { reply(callback) { ImageTransformer.apply(request) } }
    }

    override fun trimVideo(
        request: VideoTrimRequest,
        jobId: String,
        callback: (Result<MediaInfo>) -> Unit,
    ) {
        exporter.trim(request, jobId) { result ->
            mainHandler.post { callback(result.mapFailure()) }
        }
    }

    override fun cancel(jobId: String) {
        exporter.cancel(jobId)
    }

    override fun videoThumbnails(
        path: String,
        atMs: List<Long>,
        maxDimension: Long,
        callback: (Result<List<ByteArray>>) -> Unit,
    ) {
        worker.execute {
            reply(callback) { exporter.thumbnails(path, atMs, maxDimension) }
        }
    }

    // Plumbing

    /**
     * Runs [body] and posts its outcome back on the main thread, which is where
     * Pigeon's reply handlers expect to be called.
     */
    private fun <T> reply(callback: (Result<T>) -> Unit, body: () -> T) {
        val result = runCatching(body).mapFailure()
        mainHandler.post { callback(result) }
    }

    /**
     * Turns a [MonolensException] into the [FlutterError] whose code Dart
     * branches on (`monolens/cancelled` in particular). Anything else passes
     * through as-is.
     */
    private fun <T> Result<T>.mapFailure(): Result<T> {
        val error = exceptionOrNull() ?: return this
        return if (error is MonolensException) {
            Result.failure(error.toFlutterError())
        } else {
            this
        }
    }
}
